File: sols/CodeForces/1771_d2/C_Hossam_and_Trainees.py
''' C. Hossam and Trainees
https://codeforces.com/contest/1771/problem/C
'''

import io, os, sys
input = io.BytesIO(os.read(0, os.fstat(0).st_size)).readline  # decode().strip() if str
output = sys.stdout.write

def debug(*args):   
    if os.environ.get('debug') in [None, '0']: return
    from inspect import currentframe, getframeinfo
    from re import search
    frame = currentframe().f_back
    s = getframeinfo(frame).code_context[0]
    r = search(r"\((.*)\)", s).group(1)
    vnames = r.split(', ')
    var_and_vals = [f'{var}={val}' for var, val in zip(vnames, args)]
    prefix = f'{currentframe().f_back.f_lineno:02d}: '
    print(f'{prefix}{", ".join(var_and_vals)}')

from types import GeneratorType
def bootstrap(f, stack=[]):
    def wrappedfunc(*args, **kwargs):
        if stack: return f(*args, **kwargs)
        to = f(*args, **kwargs)
        while True:
            if type(to) is GeneratorType:
                stack.append(to)
                to = next(to)
            else:
                stack.pop()
                if not stack: break
                to = stack[-1].send(to)
        return to
    return wrappedfunc

class IntKeyDict(dict):
    from random import randrange
    rand = randrange(1 << 62)
    def __setitem__(self, k, v): super().__setitem__(k^self.rand, v)
    def __getitem__(self, k): return super().__getitem__(k^self.rand)
    def __contains__(self, k): return super().__contains__(k^self.rand)
    def __repr__(self): return str({k: v for k, v in self.items()})
    def get(self, k, default=None): return super().get(k^self.rand, default)
    def keys(self): return [k^self.rand for k in super().keys()]
    def items(self): return [(k^self.rand, v) for k, v in super().items()]

INF = float('inf')

# -----------------------------------------

from math import gcd 
from collections import Counter

def pollard_rho(n):
    '''return a random factor of n; O(sqrt of lpf[n])'''
    if n & 1 == 0: return 2
    if n % 3 == 0: return 3

    s = ((n - 1) & (1 - n)).bit_length() - 1
    d = n >> s
    for a in [2, 325, 9375, 28178, 450775, 9780504, 1795265022]:
        p = pow(a, d, n)
        if p == 1 or p == n - 1 or a % n == 0: continue
        for _ in range(s):
            prev = p
            p = (p * p) % n
            if p == 1: return gcd(prev - 1, n)
            if p == n - 1: break
        else:
            for i in range(2, n):
                x, y = i, (i * i + 1) % n
                f = gcd(abs(x - y), n)
                while f == 1:
                    x, y = (x * x + 1) % n, (y * y + 1) % n
                    y = (y * y + 1) % n
                    f = gcd(abs(x - y), n)
                if f != n: return f
    
    return n
    
def get_prime_counter(n, memo={}):
    '''return a Counter of the prime factorization of n'''
    if n <= 1: return Counter()
    if n in memo: return memo[n]
    f = pollard_rho(n)
    memo[n] = Counter([n]) if f == n else get_prime_counter(f, memo) + get_prime_counter(n // f, memo)
    return memo[n]


def solve(N, A):
    seen = set()
    memo = IntKeyDict()
    for a in A:
        for p in get_prime_counter(a, memo):
            if p in seen: return 'YES'
            seen.add(p)

    return 'NO'


def main():
    T = int(input())
    for _ in range(T):
        N = int(input())
        A = list(map(int, input().split()))
        res = solve(N, A)
        print(res)


if __name__ == '__main__':
    main()

File: sols/CodeForces/1771_d2/C_Hossam_and_Trainees.cpp
/* C. Hossam and Trainees
https://codeforces.com/contest/1771/problem/C
 */

// g++-12 code.cpp

#pragma GCC optimize("O3")
#pragma GCC target("avx2")

#include <bits/stdc++.h>
using namespace std;

#define FOR(i, a, b) for (int i = (a), _##i = (b); i < _##i; ++i)
#define FORD(i, a, b) for (int i = (a), _##i = (b); i >= _##i; --i)
#define REP(i, a) for (int i = 0, _##i = (a); i < _##i; ++i)
#define REPD(i, n) for (int i = (n)-1; i >= 0; --i)
#define all(x) (x).begin(), (x).end()
#define pb push_back

typedef long long LL;

vector<int> sieve(int N) {
    vector<int> primes;
    vector<int> lpf(N + 1);
    for (int i = 2; i <= N; i++) {
        if (!lpf[i]) {
            primes.pb(i);
            lpf[i] = i;
        }
        for (auto p : primes) {
            if (p * i > N || p > lpf[i]) break;
            lpf[p * i] = p;
        }
    }
    return primes;
}

vector<int> PRIMES = sieve(32000);

void solve() {
    int N, a;
    cin >> N;

    set<int> seen;
    string res = "NO";

    REP(i, N) {
        cin >> a;
        for (auto p : PRIMES) {
            if (p * p > a) break;
            if (!(a % p)) {
                if (seen.count(p)) {
                    res = "YES";
                    break;
                }
                seen.insert(p);
                while (!(a % p)) a /= p;
            }
        }
        if (a > 1) {
            if (seen.count(a)) res = "YES";
            seen.insert(a);
        }
    }

    cout << res << endl;
}

int main() {
    ios::sync_with_stdio(false);
    int T;
    cin >> T;
    for (int t = 0; t < T; t++) {
        solve();
    }
    return 0;
}
